src/data-file.ts

```diff
@@ -1,46 +1,108 @@
 import { RangeResolver } from "./resolver";
 
+interface Chunk {
+  body: string;
+  headers: { [key: string]: string };
+}
+
+export function parseMultipartBody(body: string, boundary: string): Chunk[] {
+  return body
+    .split(`--${boundary}`)
+    .reduce((chunks: Chunk[], chunk: string) => {
+      if (chunk && chunk !== "--") {
+        const [head, body] = chunk.trim().split(/\r\n\r\n/g, 2);
+        const headers = head
+          .split(/\r\n/g)
+          .reduce((headers: { [key: string]: string }, header: string) => {
+            const [key, value] = header.split(/:\s+/);
+            headers[key.toLowerCase()] = value;
+            return headers;
+          }, {});
+        chunks.push({ body, headers });
+      }
+      return chunks;
+    }, []);
+}
+
 export class DataFile {
-	private originalResolver?: RangeResolver;
-
-	private constructor(
-		private resolver: (
-			start: number,
-			end: number,
-			expectedLength?: number
-		) => Promise<ArrayBuffer>
-	) {}
-
-	static forUrl(url: string) {
-		return DataFile.forResolver(async ({ start, end }) => {
-			const response = await fetch(url, {
-				headers: { Range: `bytes=${start}-${end}` },
-			});
-			const totalLength = Number(
-				response.headers.get("Content-Range")!.split("/")[1]
-			);
-			return {
-				data: await response.arrayBuffer(),
-				totalLength: totalLength,
-			};
-		});
-	}
-
-	static forResolver(resolver: RangeResolver) {
-		const instance = new DataFile(async (start, end, expectedLength?) => {
-			const result = await resolver({ start, end, expectedLength });
-			return result.data;
-		});
-		instance.originalResolver = resolver; 
-		return instance;
+  private originalResolver?: RangeResolver;
+
+  private constructor(private resolver: RangeResolver) {}
+
+  static forUrl(url: string) {
+    return DataFile.forResolver(async (ranges) => {
+      const rangesHeader = ranges
+        .map(({ start, end }) => `${start}-${end}`)
+        .join(",");
+
+      const response = await fetch(url, {
+        headers: {
+          Range: `bytes=${rangesHeader}`,
+          Accept: "multipart/bytesranges",
+        },
+      });
+
+      if (response.status !== 206) {
+        throw new Error(`Expected 206 Partial Content, got ${response.status}`);
+      }
+
+      const contentType = response.headers.get("Content-Type");
+      if (!contentType) {
+        throw new Error("Missing Content-Type in response");
+      }
+      if (contentType.includes("multipart/byteranges")) {
+        const boundary = contentType.split("boundary=")[1];
+        const abuf = await response.arrayBuffer();
+        const text = new TextDecoder().decode(abuf);
+
+        const chunks = parseMultipartBody(text, boundary);
```

Comment:
```suggestion
        const chunks = parseMultipartBody(text, boundary.slice(0, boundary.length - 1));
```

I think this would avoid needing to pop?